sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala

```diff
@@ -137,7 +139,11 @@ abstract class IntervalNumOperation(
       val iu = IntervalUtils.getClass.getName.stripSuffix("$")
       s"""
         try {
-          ${ev.value} = $iu.$operationName($interval, $num);
+          if ($num != 0) {
+            ${ev.value} = $iu.$operationName($interval, $num);
+          } else {
+            ${ev.isNull} = true;
```

Comment:
is this needed? divide by 0 throws ArithmeticException and we return null already.

Reply:
```scala
scala> 1 / 0
java.lang.ArithmeticException: / by zero
  ... 28 elided

scala> 1 / 0.asInstanceOf[Double]
res6: Double = Infinity
```
Not so sure whether this is related

Reply:
yea, and fixing type coercion should solve the problem.

Reply:
Well,
    https://github.com/apache/spark/blob/4c53ac18221cbd630bc16352e126d6fc56143239/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala#L227

We already have a test case here which seems not able to verify this case. also cc @MaxGekk

Reply:
Can you check if your newly added tests fail if we revert the change here?

Reply:
The results you expected are in the pr description

Reply:
Can you confirm that the codegen change is necessary to fix the bug?

Reply:
```scala
[info] IntervalExpressionsSuite:
[info] - millenniums (3 seconds, 280 milliseconds)
[info] - centuries (404 milliseconds)
[info] - decades (566 milliseconds)
[info] - years (680 milliseconds)
[info] - quarters (713 milliseconds)
[info] - months (1 second, 456 milliseconds)
[info] - days (270 milliseconds)
[info] - hours (243 milliseconds)
[info] - minutes (289 milliseconds)
[info] - seconds (317 milliseconds)
[info] - milliseconds (241 milliseconds)
[info] - microseconds (247 milliseconds)
[info] - epoch (163 milliseconds)
[info] - multiply (521 milliseconds)
[info] - divide (465 milliseconds)
[info] ScalaTest
[info] Run completed in 13 seconds, 609 milliseconds.
[info] Total number of tests run: 15
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 15, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```
Add the tests with out this patch

Reply:
> Can you confirm that the codegen change is necessary to fix the bug?

Yes, I used to remove that and test all these cases I add

Reply:
Now I'm confused. What's the root cause of this bug? If the type coercion problem is fixed, why do we still hit bugs? Why the `0` becomes double?

Reply:
Can we move the 0 check to `IntervalUtils.divide`?

Reply:
yea, we must do that